### PR TITLE
Preliminary data structures for approximate queries

### PR DESCRIPTION
**Pull request by ujvl**
## What changes were proposed in this pull request?

For future use in the Atomic MultiLog:
- Thread-safe, implementation-agnostic Count Sketch
- Thread-safe, Implementation-specific Universal Sketch
- Non-thread-safe Priority Queue used for baseline comparisons
- Tests

## How was this patch tested?

Unit tests

### CODE REVIEW

File: libconfluo/confluo/container/monolog/monolog_exp2_linear.h [hunk 1]
```diff
@@ -432,8 +432,23 @@ class monolog_exp2_linear : public monolog_exp2_linear_base<T, NCONTAINERS, BUCK
   /**
    * Constructs a relaxed monolog implementation
    */
-  monolog_exp2_linear()
-      : tail_(0) {
+  monolog_exp2_linear(size_t count = 0)
```
Comment (anuragkh) [inline libconfluo/confluo/container/monolog/monolog_exp2_linear.h:435]: What is the purpose of having a monolog with initialized read-tail? The space may not be allocated for it.

File: libconfluo/confluo/container/sketch/count_min_sketch.h [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+ public:
+  typedef atomic::type<counter_t> atomic_counter_t;
+
+  /**
+   * Constructor.
+   * @param num_estimates number of estimates to track per update (depth)
+   * @param num_buckets number of buckets (width)
+   * @param manager hash manager
+   */
+  count_min_sketch(size_t b, size_t t, hash_manager manager)
```
Comment (anuragkh) [inline libconfluo/confluo/container/sketch/count_min_sketch.h:29]: Maybe switch to more descriptive variable names than `b`, `t`, `w_`, `d_`?

File: libconfluo/confluo/container/sketch/count_sketch.h [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+  count_sketch() = default;
+
+  /**
+   * Constructor.
+   * @param b number of buckets (width)
+   * @param t number of estimates per update (depth)
+   * @param m1 hash manager for buckets
+   * @param m2 hash manager for signs
+   */
+  count_sketch(size_t b, size_t t, hash_manager m1, hash_manager m2)
```
Comment (anuragkh) [inline libconfluo/confluo/container/sketch/count_sketch.h:38]: More descriptive variable names.

File: libconfluo/confluo/container/sketch/frequency_functions.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace confluo {
+namespace sketch {
+
+/**
+ * These are monotonically increasing frequency-domain functions.
+ * in f and O(f^2), where f is the frequency of the data point
+ * @tparam counter_t type of input
+ */
+template<typename arg_t = int64_t>
```
Comment (anuragkh) [inline libconfluo/confluo/container/sketch/frequency_functions.h:12]: Are these the only `f` functions we support for universal sketches? How will users provide their own (in the future)?

File: libconfluo/confluo/container/sketch/priority_queue.h [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+ /**
+  *
+  * This priority queue is not thread-safe.
+  * @tparam T type of key
+  * @tparam P type of priority
+  * @tparam E type of element to store, containing key and priority
+  */
+template<typename T, typename P, typename E = pq_element<T, P>>
+class pq : public std::priority_queue<E, std::vector<E>, std::function<bool(E, E)>> {
```
Comment (anuragkh) [inline libconfluo/confluo/container/sketch/priority_queue.h]: Perhaps name it `thread_unsafe_priority_queue`?

File: libconfluo/confluo/container/sketch/universal_sketch.h [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+#include "../../schema/record.h"
+#include "hash_manager.h"
+
+namespace confluo {
+namespace sketch {
+
+
+class universal_sketch {
+
+public:
```
Comment (anuragkh) [inline libconfluo/confluo/container/sketch/universal_sketch.h]: This indentation seems off.

File: libconfluo/confluo/container/sketch/priority_queue.h (hunk 5 above)
Comment (anuragkh) [reply libconfluo/confluo/container/sketch/priority_queue.h]: Or `thread_unsafe_pq` if that's a mouthful.

File: libconfluo/confluo/container/sketch/frequency_functions.h (hunk 4 above)
Comment (ujvl) [reply libconfluo/confluo/container/sketch/frequency_functions.h:12]: I'll add that functionality in a separate PR, wanted to keep things simple in this one

Comment (anuragkh) [reply libconfluo/confluo/container/sketch/frequency_functions.h:12]: Sounds good.

File: libconfluo/confluo/container/monolog/monolog_exp2_linear.h (hunk 1 above)
Comment (ujvl) [reply libconfluo/confluo/container/monolog/monolog_exp2_linear.h:435]: No idea, made this change a long time ago. I'll try removing it and see if it breaks anything.

**Pull request merged**